linter/linter.go

```diff
@@ -55,35 +65,38 @@ func lint(vm *jsonnet.VM, node nodeWithLocation, errWriter *ErrorWriter) {
 		return variables.FindVariables(node.node, variables.Environment{"std": &std})
 	}
 
-	variableInfo := findVariables(node)
 	for importedPath, rootNode := range roots {
 		variablesInFile[importedPath] = *findVariables(nodeWithLocation{rootNode, importedPath})
 	}
 
-	for _, v := range variableInfo.Variables {
-		if len(v.Occurences) == 0 && v.VariableKind == common.VarRegular && v.Name != "$" {
-			errWriter.writeError(vm, errors.MakeStaticError("Unused variable: "+string(v.Name), v.LocRange))
-		}
-	}
-	ec := common.ErrCollector{}
-
 	vars := make(map[string]map[ast.Node]*common.Variable)
 	for importedPath, info := range variablesInFile {
 		vars[importedPath] = info.VarAt
 	}
 
-	types.Check(node.node, roots, vars, func(currentPath, importedPath string) ast.Node {
-		node, _, err := vm.ImportAST(currentPath, importedPath)
-		if err != nil {
-			return nil
+	for _, node := range nodes {
+		variableInfo := findVariables(node)
+
+		for _, v := range variableInfo.Variables {
+			if len(v.Occurences) == 0 && v.VariableKind == common.VarRegular && v.Name != "$" {
+				errWriter.writeError(vm, errors.MakeStaticError("Unused variable: "+string(v.Name), v.LocRange))
+			}
 		}
-		return node
-	}, &ec)
+		ec := common.ErrCollector{}
+
+		types.Check(node.node, roots, vars, func(currentPath, importedPath string) ast.Node {
```

Comment:
This doesn't really take advantage of processing multiple files at once (it still builds the whole graph of type relationships each time). This is fine, though. We can do it as a separate change (which will be a performance-only change with no significant effect on the results).